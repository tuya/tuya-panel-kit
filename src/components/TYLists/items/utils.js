/* eslint-disable no-param-reassign */

/**
 * https://github.com/you-dont-need/You-Dont-Need-Lodash-Underscore#_pick
 */
export const pick = (object, keys) => {
  return keys.reduce((obj, key) => {
    if (typeof object[key] !== 'undefined') {
      obj[key] = object[key];
    }
    return obj;
  }, {});
};

export const omit = (object, keys) => {
  const shallowCopy = {
    ...object,
  };
  for (let i = 0; i < keys.length; i++) {
    const key = keys[i];
    delete shallowCopy[key];
  }
  return shallowCopy;
};
