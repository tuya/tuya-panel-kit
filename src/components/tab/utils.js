import React from 'react';

const toArray = children => {
  const arr = [];
  React.Children.forEach(children, child => {
    if (child) {
      arr.push(child);
    }
  });
  return arr;
};

const getActiveIndex = (children, activeKey) => {
  const arr = toArray(children);
  for (let i = 0; i < arr.length; i++) {
    if (arr[i].key === activeKey) {
      return i;
    }
  }
  return -1;
};

const activeKeyIsValid = (children, key) => {
  const keys = React.Children.map(children, child => child && child.key);
  return keys.indexOf(key) >= 0;
};

const getDefaultActiveKey = children => {
  let activeKey;
  React.Children.forEach(children, child => {
    if (child && !activeKey && !child.props.disabled) {
      activeKey = child.key;
    }
  });
  return activeKey;
};

export default {
  toArray,
  getActiveIndex,
  activeKeyIsValid,
  getDefaultActiveKey,
};
