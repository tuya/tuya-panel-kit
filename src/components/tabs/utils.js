import { RatioUtils, NumberUtils } from '../../utils';
import { PRESS_THRESHOLD, SWIPE_THRESHOLD, VELOCITY_THRESHOLD } from './constant';

const { winWidth } = RatioUtils;
const { inMaxMin } = NumberUtils;

/**
 * @desc 计算tab的宽度
 *
 * @param {Number} maxNum - 一屏上的tab最大数量
 */
export const getTabWidth = (maxNum = 4, tabsWidth = winWidth) => tabsWidth / maxNum;

/**
 * @desc 根据x轴的位置获取当前所处位置tab的索引
 *
 * @param {Number} deltaX - 当前 x 轴的偏移量
 * @param {Number} indexWidth - 一个索引对应的宽度
 */
export const getIndexByDeltaX = (deltaX, indexWidth) => Math.floor(deltaX / indexWidth);

/**
 * @desc 根据x轴偏移量获取距离最近tab的索引
 *
 * @param {Number} deltaX - 当前 x 轴的偏移量
 * @param {Number} indexWidth - 一个索引对应的宽度
 * @param {Number} maxIndex - 最大的索引值
 */
export const getNearestIndexByDeltaX = (deltaX, indexWidth, maxIndex = 100) => {
  if (deltaX > 0) return 0;
  return inMaxMin(0, maxIndex, Math.round(Math.abs(deltaX) / indexWidth));
};

/**
 * @desc 获取索引在循环范围内的邻近的值
 *
 * @param {Number} idx - 新索引
 * @param {Number} min - 最小值
 * @param {Number} max - 最大值
 */
export const getSiblingIndex = (index, min, max) => {
  if (index === min && index === max) {
    return [index, index];
  } else if (index === min) {
    return [index, index + 1];
  } else if (index === max) {
    return [index - 1, index];
  }
  return [index - 1, index + 1];
};

/**
 *
 * @desc 获取居中对齐的tab索引
 *
 * @param {Number} idx - 当前tab的索引
 * @param {Number} oneScreenTabNum - 一屏可放下的tab数量
 * @param {Number} totalNum - tab总数量
 */
export const getCenteredScrollIndex = (idx, oneScreenTabNum, totalNum) => {
  const min = 0;
  const max = Math.max(0, totalNum - oneScreenTabNum);
  const median = Math.ceil(oneScreenTabNum / 2) - 1;
  const scrollToIdx = idx - median;
  const index = inMaxMin(min, max, scrollToIdx);
  return index;
};

/**
 * @desc 判断是否为合法的 Press 事件
 *
 * @param {Number} dx - x轴偏移量
 * @param {Number} dy - y轴偏移量
 */
export const isValidPress = (dx, dy) =>
  Math.abs(dx) <= PRESS_THRESHOLD && Math.abs(dy) <= PRESS_THRESHOLD;

/**
 * @desc 判断是否为合法的 Swipe 事件
 *
 * @param {Number} velocity - 加速度
 * @param {Number} direction - 移动距离
 */
export const isValidSwipe = (velocity, direction) => {
  return Math.abs(velocity) > VELOCITY_THRESHOLD && Math.abs(direction) < SWIPE_THRESHOLD;
};

/**
 * @desc 根据`tabLayout`累加算出每一个`tab`的`left`偏移量
 *
 * @param {Array} tabLayout - tab的布局属性，包含 x, y, width, height
 */
export const reduceTabLayoutLeft = tabLayout => {
  return tabLayout.reduce((acc, cur, index) => {
    if (index === 0) {
      return [{ ...cur, left: cur.x }];
    }
    const prev = acc[index - 1];
    const prevTotalX = prev.left + prev.width + prev.x;
    return [
      ...acc,
      {
        ...cur,
        left: prevTotalX + cur.x,
      },
    ];
  }, {});
};
