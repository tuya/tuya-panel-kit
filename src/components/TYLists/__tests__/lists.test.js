import { View, Text, Image } from 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import Lists from '../lists';
import { ThemeUtils } from '../../../utils';

const { ThemeProvider } = ThemeUtils;

const Res = {
  color: require('../../color-picker/color-picker.png'),
};

describe('Lists Component', () => {
  it('basic render', () => {
    const sections = [1].map(v => ({
      title: 'Section Title',
      footer: true,
      data: [1].map(val => ({
        styles: {
          container: { height: 64 },
          content: { backgroundColor: 'red' },
          contentLeft: { width: 48, height: 48 },
          contentCenter: { backgroundColor: 'blue' },
          contentRight: { width: 48, height: 48 },
          title: { color: 'yellow' },
          subTitle: { color: '#000' },
          valueText: { color: 'pink' },
        },
        title: `title_${v}`,
        subTitle: `subTitle_${v}`,
        Icon: <Image style={{}} source={Res.color} />,
        onPress: jest.fn(),
        renderItem: () => <View />,
      })),
    }));
    const component = renderer.create(<Lists sections={sections} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('basic render 1', () => {
    const sections = [1].map(v => ({
      title: 'Section Title',
      footer: true,
      data: [1, 2].map((val, index) => ({
        styles: {
          container: { height: 64 },
          content: { backgroundColor: 'red' },
          contentLeft: { width: 48, height: 48 },
          contentCenter: { backgroundColor: 'blue' },
          contentRight: { width: 48, height: 48 },
          title: { color: 'yellow' },
          subTitle: { color: '#000' },
          valueText: { color: 'pink' },
        },
        title: `title_${v}`,
        subTitle: `subTitle_${v}`,
        Icon: <Image style={{}} source={Res.color} />,
        onPress: jest.fn(),
        value: index === 0 ? true : undefined,
      })),
    }));
    const component = renderer.create(
      <ThemeProvider>
        <Lists sections={sections} />
      </ThemeProvider>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should not render SectionHeader without title', () => {
    const sections = [
      {
        title: null,
        data: [],
      },
    ];
    const component = renderer.create(<Lists sections={sections} ListHeaderComponent={<View />} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should able to use SectionList props', () => {
    const sections = [
      {
        title: 'Section Title',
        data: [1].map(v => ({
          key: v,
          title: `title_${v}`,
          value: `${v}`,
          Action: () => (
            <View>
              <Text>Action</Text>
            </View>
          ),
        })),
      },
    ];
    const component = renderer.create(<Lists sections={sections} ListHeaderComponent={<View />} />);
    expect(component.toJSON()).toMatchSnapshot();
  });
});
