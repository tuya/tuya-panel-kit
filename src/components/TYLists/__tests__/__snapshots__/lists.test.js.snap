// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lists Component basic render 1`] = `
<RCTScrollView
  contentContainerStyle={
    Array [
      Object {
        "backgroundColor": "#f8f8f8",
      },
      undefined,
    ]
  }
  data={
    Array [
      Object {
        "data": Array [
          Object {
            "Icon": <Image
              source={1}
              style={Object {}}
            />,
            "onPress": [MockFunction],
            "renderItem": [Function],
            "styles": Object {
              "container": Object {
                "height": 64,
              },
              "content": Object {
                "backgroundColor": "red",
              },
              "contentCenter": Object {
                "backgroundColor": "blue",
              },
              "contentLeft": Object {
                "height": 48,
                "width": 48,
              },
              "contentRight": Object {
                "height": 48,
                "width": 48,
              },
              "subTitle": Object {
                "color": "#000",
              },
              "title": Object {
                "color": "yellow",
              },
              "valueText": Object {
                "color": "pink",
              },
            },
            "subTitle": "subTitle_1",
            "title": "title_1",
          },
        ],
        "footer": true,
        "title": "Section Title",
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  headerStyle={null}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  renderItem={[Function]}
  renderSectionFooter={[Function]}
  renderSectionHeader={[Function]}
  scrollEventThrottle={50}
  sections={
    Array [
      Object {
        "data": Array [
          Object {
            "Icon": <Image
              source={1}
              style={Object {}}
            />,
            "onPress": [MockFunction],
            "renderItem": [Function],
            "styles": Object {
              "container": Object {
                "height": 64,
              },
              "content": Object {
                "backgroundColor": "red",
              },
              "contentCenter": Object {
                "backgroundColor": "blue",
              },
              "contentLeft": Object {
                "height": 48,
                "width": 48,
              },
              "contentRight": Object {
                "height": 48,
                "width": 48,
              },
              "subTitle": Object {
                "color": "#000",
              },
              "title": Object {
                "color": "yellow",
              },
              "valueText": Object {
                "color": "pink",
              },
            },
            "subTitle": "subTitle_1",
            "title": "title_1",
          },
        ],
        "footer": true,
        "title": "Section Title",
      },
    ]
  }
  showsVerticalScrollIndicator={false}
  stickyHeaderIndices={Array []}
  stickySectionHeadersEnabled={false}
  updateCellsBatchingPeriod={50}
  useART={false}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Array [
            Object {
              "marginBottom": 8,
              "marginLeft": 32,
              "marginRight": 32,
              "marginTop": 24,
            },
            Array [
              Object {
                "marginTop": 24,
              },
              null,
            ],
          ]
        }
      >
        <Text
          accessible={true}
          align={null}
          allowFontScaling={false}
          color={null}
          ellipsizeMode="tail"
          size="normal"
          style={
            Array [
              Object {
                "backgroundColor": "transparent",
                "color": "#333",
                "fontSize": 17,
                "lineHeight": 24,
              },
              Array [
                Object {
                  "color": "rgba(51, 51, 51, 0.5)",
                },
                undefined,
              ],
            ]
          }
          type="paragraph"
          weight={null}
        >
          Section Title
        </Text>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View />
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Array [
            Object {
              "marginBottom": 24,
              "marginLeft": 32,
              "marginRight": 32,
              "marginTop": 8,
            },
            undefined,
          ]
        }
      >
        <Text
          accessible={true}
          align={null}
          allowFontScaling={false}
          color={null}
          ellipsizeMode="tail"
          size="normal"
          style={
            Array [
              Object {
                "backgroundColor": "transparent",
                "color": "#333",
                "fontSize": 17,
                "lineHeight": 24,
              },
              Array [
                Object {
                  "color": "rgba(51, 51, 51, 0.5)",
                },
                undefined,
              ],
            ]
          }
          type="paragraph"
          weight={null}
        />
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`Lists Component basic render 1 1`] = `
<RCTScrollView
  contentContainerStyle={
    Array [
      Object {
        "backgroundColor": "#f8f8f8",
      },
      undefined,
    ]
  }
  data={
    Array [
      Object {
        "data": Array [
          Object {
            "Icon": <Image
              source={1}
              style={Object {}}
            />,
            "onPress": [MockFunction],
            "styles": Object {
              "container": Object {
                "height": 64,
              },
              "content": Object {
                "backgroundColor": "red",
              },
              "contentCenter": Object {
                "backgroundColor": "blue",
              },
              "contentLeft": Object {
                "height": 48,
                "width": 48,
              },
              "contentRight": Object {
                "height": 48,
                "width": 48,
              },
              "subTitle": Object {
                "color": "#000",
              },
              "title": Object {
                "color": "yellow",
              },
              "valueText": Object {
                "color": "pink",
              },
            },
            "subTitle": "subTitle_1",
            "title": "title_1",
            "value": true,
          },
          Object {
            "Icon": <Image
              source={1}
              style={Object {}}
            />,
            "onPress": [MockFunction],
            "styles": Object {
              "container": Object {
                "height": 64,
              },
              "content": Object {
                "backgroundColor": "red",
              },
              "contentCenter": Object {
                "backgroundColor": "blue",
              },
              "contentLeft": Object {
                "height": 48,
                "width": 48,
              },
              "contentRight": Object {
                "height": 48,
                "width": 48,
              },
              "subTitle": Object {
                "color": "#000",
              },
              "title": Object {
                "color": "yellow",
              },
              "valueText": Object {
                "color": "pink",
              },
            },
            "subTitle": "subTitle_1",
            "title": "title_1",
            "value": undefined,
          },
        ],
        "footer": true,
        "title": "Section Title",
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  headerStyle={null}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  renderItem={[Function]}
  renderSectionFooter={[Function]}
  renderSectionHeader={[Function]}
  scrollEventThrottle={50}
  sections={
    Array [
      Object {
        "data": Array [
          Object {
            "Icon": <Image
              source={1}
              style={Object {}}
            />,
            "onPress": [MockFunction],
            "styles": Object {
              "container": Object {
                "height": 64,
              },
              "content": Object {
                "backgroundColor": "red",
              },
              "contentCenter": Object {
                "backgroundColor": "blue",
              },
              "contentLeft": Object {
                "height": 48,
                "width": 48,
              },
              "contentRight": Object {
                "height": 48,
                "width": 48,
              },
              "subTitle": Object {
                "color": "#000",
              },
              "title": Object {
                "color": "yellow",
              },
              "valueText": Object {
                "color": "pink",
              },
            },
            "subTitle": "subTitle_1",
            "title": "title_1",
            "value": true,
          },
          Object {
            "Icon": <Image
              source={1}
              style={Object {}}
            />,
            "onPress": [MockFunction],
            "styles": Object {
              "container": Object {
                "height": 64,
              },
              "content": Object {
                "backgroundColor": "red",
              },
              "contentCenter": Object {
                "backgroundColor": "blue",
              },
              "contentLeft": Object {
                "height": 48,
                "width": 48,
              },
              "contentRight": Object {
                "height": 48,
                "width": 48,
              },
              "subTitle": Object {
                "color": "#000",
              },
              "title": Object {
                "color": "yellow",
              },
              "valueText": Object {
                "color": "pink",
              },
            },
            "subTitle": "subTitle_1",
            "title": "title_1",
            "value": undefined,
          },
        ],
        "footer": true,
        "title": "Section Title",
      },
    ]
  }
  showsVerticalScrollIndicator={false}
  stickyHeaderIndices={Array []}
  stickySectionHeadersEnabled={false}
  updateCellsBatchingPeriod={50}
  useART={false}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Array [
            Object {
              "marginBottom": 8,
              "marginLeft": 32,
              "marginRight": 32,
              "marginTop": 24,
            },
            Array [
              Object {
                "marginTop": 24,
              },
              null,
            ],
          ]
        }
      >
        <Text
          accessible={true}
          align={null}
          allowFontScaling={false}
          color={null}
          ellipsizeMode="tail"
          size="normal"
          style={
            Array [
              Object {
                "backgroundColor": "transparent",
                "color": "#333",
                "fontSize": 17,
                "lineHeight": 24,
              },
              Array [
                Object {
                  "color": "rgba(51, 51, 51, 0.5)",
                },
                undefined,
              ],
            ]
          }
          type="paragraph"
          weight={null}
        >
          Section Title
        </Text>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View>
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignSelf": "stretch",
              "backgroundColor": "#fff",
              "borderRadius": 0,
              "flexGrow": 0,
              "height": 64,
              "marginBottom": 0,
              "marginLeft": 0,
              "marginRight": 0,
              "marginTop": 0,
              "minHeight": 48,
              "opacity": 1,
            }
          }
        >
          <View
            disabled={false}
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "borderRadius": 0,
                  "flexDirection": "row",
                  "opacity": 1,
                  "paddingBottom": 12,
                  "paddingLeft": 28,
                  "paddingRight": 28,
                  "paddingTop": 12,
                },
                Array [
                  false,
                  undefined,
                  Object {
                    "backgroundColor": "red",
                  },
                ],
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "justifyContent": "center",
                    "marginRight": 20,
                  },
                  Object {
                    "height": 48,
                    "width": 48,
                  },
                ]
              }
            >
              <Image
                source={1}
                style={Object {}}
              />
            </View>
            <View
              style={
                Array [
                  Object {
                    "flexBasis": 0,
                    "flexGrow": 1,
                    "flexShrink": 1,
                    "justifyContent": "center",
                  },
                  Object {
                    "backgroundColor": "blue",
                  },
                ]
              }
            >
              <Text
                accessible={true}
                align={null}
                allowFontScaling={false}
                color={null}
                ellipsizeMode="tail"
                size="small"
                style={
                  Array [
                    Object {
                      "backgroundColor": "transparent",
                      "color": "#333",
                      "fontSize": 17,
                      "lineHeight": 24,
                    },
                    Array [
                      Object {
                        "color": "#333",
                      },
                      Array [
                        undefined,
                        Object {
                          "color": "yellow",
                        },
                      ],
                    ],
                  ]
                }
                type="title"
                weight={null}
              >
                title_1
              </Text>
              <Text
                accessible={true}
                align={null}
                allowFontScaling={false}
                color={null}
                ellipsizeMode="tail"
                size="large"
                style={
                  Array [
                    Object {
                      "backgroundColor": "transparent",
                      "color": "#333",
                      "fontSize": 17,
                      "lineHeight": 24,
                    },
                    Array [
                      Object {
                        "color": "rgba(51, 51, 51, 0.5)",
                        "marginTop": 4,
                      },
                      Array [
                        undefined,
                        Object {
                          "color": "#000",
                        },
                      ],
                    ],
                  ]
                }
                type="paragraph"
                weight={null}
              >
                subTitle_1
              </Text>
            </View>
            <View
              accessible={true}
              isTVSelectable={true}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "height": 48,
                  "marginLeft": 20,
                  "opacity": 1,
                  "width": 48,
                }
              }
            >
              <View
                needsOffscreenAlphaCompositing={true}
                style={
                  Array [
                    undefined,
                    false,
                  ]
                }
              >
                <View
                  accessibilityLabel="SwitchButton"
                  accessible={true}
                  isTVSelectable={true}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "transparent",
                      "height": NaN,
                      "justifyContent": "center",
                      "opacity": 1,
                      "width": NaN,
                    }
                  }
                >
                  <View
                    style={
                      Array [
                        Object {
                          "borderRadius": NaN,
                          "height": undefined,
                          "justifyContent": "center",
                          "width": undefined,
                        },
                        Object {
                          "backgroundColor": "#44DB5E",
                          "borderColor": "#44DB5E",
                        },
                      ]
                    }
                  />
                  <View
                    style={
                      Object {
                        "alignItems": "center",
                        "alignSelf": "flex-start",
                        "backgroundColor": "#fff",
                        "borderRadius": NaN,
                        "elevation": 2,
                        "height": undefined,
                        "justifyContent": "center",
                        "position": "absolute",
                        "shadowColor": "rgba(0, 0, 0, 0.3)",
                        "shadowOffset": Object {
                          "height": 3,
                          "width": 0,
                        },
                        "shadowOpacity": 1,
                        "shadowRadius": 3,
                        "transform": Array [
                          Object {
                            "translateX": NaN,
                          },
                        ],
                        "width": undefined,
                      }
                    }
                  />
                </View>
              </View>
            </View>
          </View>
        </View>
        <View
          style={
            Array [
              Object {
                "backgroundColor": "rgba(51, 51, 51, 0.1)",
                "height": 0.5,
                "marginLeft": 32,
              },
              null,
            ]
          }
        />
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignSelf": "stretch",
            "backgroundColor": "#fff",
            "borderRadius": 0,
            "flexGrow": 0,
            "height": 64,
            "marginBottom": 0,
            "marginLeft": 0,
            "marginRight": 0,
            "marginTop": 0,
            "minHeight": 48,
            "opacity": 1,
          }
        }
      >
        <View
          disabled={false}
          style={
            Array [
              Object {
                "alignItems": "center",
                "borderRadius": 0,
                "flexDirection": "row",
                "opacity": 1,
                "paddingBottom": 12,
                "paddingLeft": 28,
                "paddingRight": 28,
                "paddingTop": 12,
              },
              Array [
                false,
                undefined,
                Object {
                  "backgroundColor": "red",
                },
              ],
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "justifyContent": "center",
                  "marginRight": 20,
                },
                Object {
                  "height": 48,
                  "width": 48,
                },
              ]
            }
          >
            <Image
              source={1}
              style={Object {}}
            />
          </View>
          <View
            style={
              Array [
                Object {
                  "flexBasis": 0,
                  "flexGrow": 1,
                  "flexShrink": 1,
                  "justifyContent": "center",
                },
                Object {
                  "backgroundColor": "blue",
                },
              ]
            }
          >
            <Text
              accessible={true}
              align={null}
              allowFontScaling={false}
              color={null}
              ellipsizeMode="tail"
              size="small"
              style={
                Array [
                  Object {
                    "backgroundColor": "transparent",
                    "color": "#333",
                    "fontSize": 17,
                    "lineHeight": 24,
                  },
                  Array [
                    Object {
                      "color": "#333",
                    },
                    Array [
                      undefined,
                      Object {
                        "color": "yellow",
                      },
                    ],
                  ],
                ]
              }
              type="title"
              weight={null}
            >
              title_1
            </Text>
            <Text
              accessible={true}
              align={null}
              allowFontScaling={false}
              color={null}
              ellipsizeMode="tail"
              size="large"
              style={
                Array [
                  Object {
                    "backgroundColor": "transparent",
                    "color": "#333",
                    "fontSize": 17,
                    "lineHeight": 24,
                  },
                  Array [
                    Object {
                      "color": "rgba(51, 51, 51, 0.5)",
                      "marginTop": 4,
                    },
                    Array [
                      undefined,
                      Object {
                        "color": "#000",
                      },
                    ],
                  ],
                ]
              }
              type="paragraph"
              weight={null}
            >
              subTitle_1
            </Text>
          </View>
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "height": 48,
                "marginLeft": 20,
                "opacity": 1,
                "width": 48,
              }
            }
          />
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Array [
            Object {
              "marginBottom": 24,
              "marginLeft": 32,
              "marginRight": 32,
              "marginTop": 8,
            },
            undefined,
          ]
        }
      >
        <Text
          accessible={true}
          align={null}
          allowFontScaling={false}
          color={null}
          ellipsizeMode="tail"
          size="normal"
          style={
            Array [
              Object {
                "backgroundColor": "transparent",
                "color": "#333",
                "fontSize": 17,
                "lineHeight": 24,
              },
              Array [
                Object {
                  "color": "rgba(51, 51, 51, 0.5)",
                },
                undefined,
              ],
            ]
          }
          type="paragraph"
          weight={null}
        />
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`Lists Component should able to use SectionList props 1`] = `
<RCTScrollView
  ListHeaderComponent={<View />}
  contentContainerStyle={
    Array [
      Object {
        "backgroundColor": "#f8f8f8",
      },
      undefined,
    ]
  }
  data={
    Array [
      Object {
        "data": Array [
          Object {
            "Action": [Function],
            "key": 1,
            "title": "title_1",
            "value": "1",
          },
        ],
        "title": "Section Title",
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  headerStyle={null}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  renderItem={[Function]}
  renderSectionFooter={[Function]}
  renderSectionHeader={[Function]}
  scrollEventThrottle={50}
  sections={
    Array [
      Object {
        "data": Array [
          Object {
            "Action": [Function],
            "key": 1,
            "title": "title_1",
            "value": "1",
          },
        ],
        "title": "Section Title",
      },
    ]
  }
  showsVerticalScrollIndicator={false}
  stickyHeaderIndices={Array []}
  stickySectionHeadersEnabled={false}
  updateCellsBatchingPeriod={50}
  useART={false}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View />
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Array [
            Object {
              "marginBottom": 8,
              "marginLeft": 32,
              "marginRight": 32,
              "marginTop": 24,
            },
            Array [
              Object {
                "marginTop": 24,
              },
              null,
            ],
          ]
        }
      >
        <Text
          accessible={true}
          align={null}
          allowFontScaling={false}
          color={null}
          ellipsizeMode="tail"
          size="normal"
          style={
            Array [
              Object {
                "backgroundColor": "transparent",
                "color": "#333",
                "fontSize": 17,
                "lineHeight": 24,
              },
              Array [
                Object {
                  "color": "rgba(51, 51, 51, 0.5)",
                },
                undefined,
              ],
            ]
          }
          type="paragraph"
          weight={null}
        >
          Section Title
        </Text>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignSelf": "stretch",
            "backgroundColor": "#fff",
            "borderRadius": 0,
            "flexGrow": 0,
            "marginBottom": 0,
            "marginLeft": 0,
            "marginRight": 0,
            "marginTop": 0,
            "minHeight": 48,
            "opacity": 1,
          }
        }
      >
        <View
          disabled={false}
          style={
            Array [
              Object {
                "alignItems": "center",
                "borderRadius": 0,
                "flexDirection": "row",
                "opacity": 1,
                "paddingBottom": 12,
                "paddingLeft": 28,
                "paddingRight": 28,
                "paddingTop": 12,
              },
              Array [
                false,
                undefined,
                undefined,
              ],
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "flexBasis": 0,
                  "flexGrow": 1,
                  "flexShrink": 1,
                  "justifyContent": "center",
                },
                undefined,
              ]
            }
          >
            <Text
              accessible={true}
              align={null}
              allowFontScaling={false}
              color={null}
              ellipsizeMode="tail"
              size="small"
              style={
                Array [
                  Object {
                    "backgroundColor": "transparent",
                    "color": "#333",
                    "fontSize": 17,
                    "lineHeight": 24,
                  },
                  Array [
                    Object {
                      "color": "#333",
                    },
                    Array [
                      undefined,
                      undefined,
                    ],
                  ],
                ]
              }
              type="title"
              weight={null}
            >
              title_1
            </Text>
          </View>
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "marginLeft": 20,
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                }
              }
            >
              <Text
                accessible={true}
                align={null}
                allowFontScaling={false}
                color={null}
                ellipsizeMode="tail"
                size="large"
                style={
                  Array [
                    Object {
                      "backgroundColor": "transparent",
                      "color": "#333",
                      "fontSize": 17,
                      "lineHeight": 24,
                    },
                    Array [
                      Object {
                        "color": "rgba(51, 51, 51, 0.5)",
                      },
                      Array [
                        undefined,
                        undefined,
                      ],
                    ],
                  ]
                }
                type="paragraph"
                weight={null}
              >
                1
              </Text>
              <View>
                <Text
                  accessible={true}
                  allowFontScaling={true}
                  ellipsizeMode="tail"
                >
                  Action
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    />
  </View>
</RCTScrollView>
`;

exports[`Lists Component should not render SectionHeader without title 1`] = `
<RCTScrollView
  ListHeaderComponent={<View />}
  contentContainerStyle={
    Array [
      Object {
        "backgroundColor": "#f8f8f8",
      },
      undefined,
    ]
  }
  data={
    Array [
      Object {
        "data": Array [],
        "title": null,
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  headerStyle={null}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  renderItem={[Function]}
  renderSectionFooter={[Function]}
  renderSectionHeader={[Function]}
  scrollEventThrottle={50}
  sections={
    Array [
      Object {
        "data": Array [],
        "title": null,
      },
    ]
  }
  showsVerticalScrollIndicator={false}
  stickyHeaderIndices={Array []}
  stickySectionHeadersEnabled={false}
  updateCellsBatchingPeriod={50}
  useART={false}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View />
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Object {
            "marginTop": 12,
          }
        }
      />
    </View>
    <View
      onLayout={[Function]}
      style={null}
    />
  </View>
</RCTScrollView>
`;
